src/linalg/basic/matrix.rs

```diff
@@ -49,27 +51,75 @@ pub struct DenseMatrixMutView<'a, T: Debug + Display + Copy + Sized> {
     column_major: bool,
 }
 
+// functional utility functions used across types
+fn is_valid_matrix_window(
+    mrows: usize,
+    mcols: usize,
+    vrows: &Range<usize>,
+    vcols: &Range<usize>,
+) -> bool {
+    debug_assert!(
+        vrows.end <= mrows && vcols.end <= mcols,
+        "The window end is outside of the matrix range"
+    );
+    debug_assert!(
+        vrows.start <= mrows && vcols.start <= mcols,
+        "The window start is outside of the matrix range"
+    );
+    debug_assert!(
+        // depends on a properly formed range
+        vrows.start <= vrows.end && vcols.start <= vcols.end,
+        "Invalid range: start <= end failed"
+    );
+
+    !(vrows.end <= mrows && vcols.end <= mcols && vrows.start <= mrows && vcols.start <= mcols)
+}
+fn start_end_stride(
+    mrows: usize,
+    mcols: usize,
+    vrows: &Range<usize>,
+    vcols: &Range<usize>,
+    column_major: bool,
+) -> (usize, usize, usize) {
+    let (start, end, stride) = if column_major {
+        (
+            vrows.start + vcols.start * mrows,
+            vrows.end + (vcols.end - 1) * mrows,
+            mrows,
+        )
+    } else {
+        (
+            vrows.start * mcols + vcols.start,
+            (vrows.end - 1) * mcols + vcols.end,
+            mcols,
+        )
+    };
+    (start, end, stride)
+}
+
 impl<'a, T: Debug + Display + Copy + Sized> DenseMatrixView<'a, T> {
-    fn new(m: &'a DenseMatrix<T>, rows: Range<usize>, cols: Range<usize>) -> Self {
-        let (start, end, stride) = if m.column_major {
-            (
-                rows.start + cols.start * m.nrows,
-                rows.end + (cols.end - 1) * m.nrows,
-                m.nrows,
-            )
+    fn new(
+        m: &'a DenseMatrix<T>,
+        vrows: Range<usize>,
+        vcols: Range<usize>,
+    ) -> Result<Self, Failed> {
+        let (mrows, mcols) = m.shape();
+
+        if is_valid_matrix_window(mrows, mcols, &vrows, &vcols) {
+            Err(Failed::input(&format!(
```

Comment:
There are some linting issues like this across the code, could you fix it
```
useless use of `format!`
```

Reply:
I saw that. I kept it "as is" for now for 2 reasons:
1. I replicated how the errors were generated elsewhere  (e.g., src/decomposition/svd.rs
```rust
        if parameters.n_components >= p {
            return Err(Failed::fit(&format!(
                "Number of components, n_components should be < number of attributes ({p})"
            )));
        }
```

2. using `format` is preferred when it includes a variable (~"string literal") to help inform the error.  I don't have a variable for now.  When we figure out the most useful message, it's easier to change.